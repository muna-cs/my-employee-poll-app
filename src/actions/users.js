export const RECEIVE_USERS = "RECEIVE_USERS";
export const SAVE_ANSWER="SAVE_ANSWER";
export const SAVE_QUESTION="SAVE_QUESTION";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function saveAnswer(authedUser, qid, answer) {
  return {
    type: SAVE_ANSWER,
    authedUser,
    qid,
    answer,
  };

};

export function saveQuestion(question) {
  return {
    type: SAVE_QUESTION,
    question
  };
};