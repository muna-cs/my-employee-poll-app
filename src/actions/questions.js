import {saveNewQuestion, saveNewQuestionAnswer }from '../utils/api'
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { saveAnswer , saveQuestion } from './users';
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER = "ADD_ANSWER";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addAnswer(authedUser,qid,answer) {
  return {
    type: ADD_ANSWER,
    authedUser, qid, answer
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question, 
  };
}

export function handleAddAnswer(authedUser,qid,answer) {
  return (dispatch) => {
    dispatch(addAnswer(authedUser, qid, answer));
    dispatch(saveAnswer(authedUser, qid, answer));
    return saveNewQuestionAnswer({authedUser:authedUser, qid, answer})
    .catch((e) => {
      console.warn("Error in Add Answer: ", e);
      dispatch(addAnswer(authedUser, qid, answer));
      dispatch(saveAnswer(authedUser, qid, answer));
    });
       
      
};
}


export function handleAddQuestion(authedUser,optionOneText,optionTwoText) {
  return (dispatch) => {
    dispatch(showLoading());
    return saveNewQuestion({ optionOneText, optionTwoText,author:authedUser})
       .then((q) => {
       dispatch(addQuestion(q));
       dispatch(saveQuestion(q));
        }).then(() => dispatch(hideLoading()));
};
}