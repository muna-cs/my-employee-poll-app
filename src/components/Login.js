import {Button, Image, Row,Col,Alert} from 'react-bootstrap';
import { connect } from "react-redux";
import logo from '../asset/login.jpg'
import { setAuthedUser } from '../actions/authedUser';
import { useNavigate } from 'react-router-dom';
import React, { useState } from 'react'
const Login= ( props ) =>{
 const [msgError, setMsg]= useState(false);
  let navigate = useNavigate();
  
  const handelUserLogin=(e)=> {
  e.preventDefault();
  const { dispatch } = props;
  const username= document.getElementById("UserName").value;
  const pass= document.getElementById("Password").value;
  const currentuser= props.usersInfo.filter(user=> user.id===username && user.password===pass);
  if(currentuser.length>0)
  {
  dispatch(setAuthedUser( currentuser[0].id));
  }
  else
  setMsg(true);
}

return(

    <div className='text-center m-5'>
      <h1 >LOGIN </h1>
      <Row>
      <Col sm></Col>
      <Col sm>
      <Image src={logo} alt="Logo" />
     </Col>
     <Col sm></Col>
     </Row>
     <Row>
      {msgError?
      <Alert data-testid="ErrorAlert" variant="danger">User is not Exist! Try Again</Alert>:""}
     </Row>
     <Row className="mt-3">
      <Col sm></Col>
      <Col sm>
      <label className="h4">UserName</label>
      <input
                            type="text"
                            name="UserName"
                            id="UserName"
                            className="form-control" data-testid="UserName"
                            ></input>
      </Col>
      <Col sm></Col>
     </Row>
     <Row className="mt-3">
      <Col sm></Col>
      <Col sm>
      <label className="h4">Password</label>
      <input
                            type="text"
                            name="Password"
                            id="Password"
                            className="form-control" data-testid="Password"
                            ></input>
      </Col>
      <Col sm></Col>
     </Row>
     <Row className=' mt-3'>
      <Col sm></Col>
      <Col sm><Button variant="primary" size="lg" data-testid="LoginButton"  onClick={handelUserLogin}>Login</Button></Col>
      <Col sm></Col>
      </Row>
      </div>

)

}
const mapStateToProps = ({ users }) => ({
  usersInfo: Object.values(users)
});
export default connect(mapStateToProps)(Login);