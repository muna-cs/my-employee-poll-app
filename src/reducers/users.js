import { RECEIVE_USERS ,SAVE_ANSWER, SAVE_QUESTION } from "../actions/users";

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
      case SAVE_ANSWER:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
            
          },
        },
      };
      case SAVE_QUESTION:
      return {
        ...state,
        [action.question.author]: {
          ...state[action.question.author],
          questions: [action.question.id].concat(state[action.question.author].questions),
        },
      };
    default:
      return state;
  }
}