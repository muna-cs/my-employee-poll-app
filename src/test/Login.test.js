import Login from '../components/Login' 
import * as React from 'react';
import { render ,fireEvent} from '@testing-library/react';
import { MemoryRouter as Router,BrowserRouter } from 'react-router-dom';
import { Provider } from "react-redux";
import { configureStore } from '@reduxjs/toolkit';
import authedUser from '../reducers/authedUser';
import users from '../reducers/users';
import questions from '../reducers/questions';
import logger from '../middleware/logger';
import thunk from 'redux-thunk';

export const store = configureStore({
  reducer: {
    authedUser,
    users,
    questions
  },
  middleware: [thunk, logger]
});
 



describe('Login', () => {
    var component =  render(
        <Provider store={store}>
    <Router>
       <Login/></Router>
       </Provider>
        );
   
        it('will create create snapshot', () => {
       
            expect(component).toMatchSnapshot();
        });
    });


    it('will display an error when expected fields not exist', () => {
        var component =  render(
            <Provider store={store}>
        <Router>
           <Login/></Router>
           </Provider>
            );
        var usernameInput = component.getByTestId('UserName')
        var passwordInput = component.getByTestId('Password')
        expect(usernameInput).toBeInTheDocument();
        expect(passwordInput).toBeInTheDocument();

        var LoginButton = component.getByText('Login')
        expect(LoginButton).toBeInTheDocument();
        

         });

         it('will display an alert if username or password is wrong', () => {
            var component =  render(
                <Provider store={store}>
            <Router>
               <Login/></Router>
               </Provider>
                );
    
              
            var LoginButton = component.getByTestId('LoginButton');
            fireEvent.click(LoginButton);
            expect(component.getByTestId('ErrorAlert')).toBeInTheDocument();
        });