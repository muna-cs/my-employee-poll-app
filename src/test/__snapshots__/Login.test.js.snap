// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will create create snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="text-center m-5"
      >
        <h1>
          LOGIN 
        </h1>
        <div
          class="row"
        >
          <div
            class="col-sm"
          />
          <div
            class="col-sm"
          >
            <img
              alt="Logo"
              class=""
              src="login.jpg"
            />
          </div>
          <div
            class="col-sm"
          />
        </div>
        <div
          class="row"
        />
        <div
          class="mt-3 row"
        >
          <div
            class="col-sm"
          />
          <div
            class="col-sm"
          >
            <label
              class="h4"
            >
              UserName
            </label>
            <input
              class="form-control"
              data-testid="UserName"
              id="UserName"
              name="UserName"
              type="text"
            />
          </div>
          <div
            class="col-sm"
          />
        </div>
        <div
          class="mt-3 row"
        >
          <div
            class="col-sm"
          />
          <div
            class="col-sm"
          >
            <label
              class="h4"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="Password"
              id="Password"
              name="Password"
              type="text"
            />
          </div>
          <div
            class="col-sm"
          />
        </div>
        <div
          class=" mt-3 row"
        >
          <div
            class="col-sm"
          />
          <div
            class="col-sm"
          >
            <button
              class="btn btn-primary btn-lg"
              data-testid="LoginButton"
              type="button"
            >
              Login
            </button>
          </div>
          <div
            class="col-sm"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="text-center m-5"
    >
      <h1>
        LOGIN 
      </h1>
      <div
        class="row"
      >
        <div
          class="col-sm"
        />
        <div
          class="col-sm"
        >
          <img
            alt="Logo"
            class=""
            src="login.jpg"
          />
        </div>
        <div
          class="col-sm"
        />
      </div>
      <div
        class="row"
      />
      <div
        class="mt-3 row"
      >
        <div
          class="col-sm"
        />
        <div
          class="col-sm"
        >
          <label
            class="h4"
          >
            UserName
          </label>
          <input
            class="form-control"
            data-testid="UserName"
            id="UserName"
            name="UserName"
            type="text"
          />
        </div>
        <div
          class="col-sm"
        />
      </div>
      <div
        class="mt-3 row"
      >
        <div
          class="col-sm"
        />
        <div
          class="col-sm"
        >
          <label
            class="h4"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="Password"
            id="Password"
            name="Password"
            type="text"
          />
        </div>
        <div
          class="col-sm"
        />
      </div>
      <div
        class=" mt-3 row"
      >
        <div
          class="col-sm"
        />
        <div
          class="col-sm"
        >
          <button
            class="btn btn-primary btn-lg"
            data-testid="LoginButton"
            type="button"
          >
            Login
          </button>
        </div>
        <div
          class="col-sm"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
